lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/Versions.java

```diff
@@ -0,0 +1,73 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.base;
+
+import io.trino.spi.connector.ConnectorContext;
+import io.trino.spi.connector.ConnectorFactory;
+
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public final class Versions
+{
+    private Versions() {}
+
+    /**
+     * Check if the SPI version of the Trino server matches exactly the SPI version the connector plugin was built for.
+     * Using plugins built for a different version of Trino may fail at runtime, especially if plugin author
+     * chooses not to maintain compatibility with older SPI versions, as happens for plugins maintained together with
+     * the Trino project.
+     */
+    public static void checkSpiVersion(ConnectorContext context, ConnectorFactory connectorFactory)
+    {
+        String spiVersion = context.getSpiVersion();
+        Optional<String> pluginVersion = getPluginMavenVersion(connectorFactory);
+
+        if (pluginVersion.isEmpty()) {
+            // Assume we're in tests. In tests, plugin version is unknown and SPI version may be known, e.g. when running single module's tests from maven.
+            return;
+        }
+
+        checkState(
+                spiVersion.equals(pluginVersion.get()),
+                format(
+                        "Trino SPI version %s does not match %s connector version %s. The connector cannot be used with SPI version other than it was compiled for.",
+                        spiVersion,
+                        connectorFactory.getName(),
+                        pluginVersion.get()));
+    }
+
+    private static Optional<String> getPluginMavenVersion(ConnectorFactory connectorFactory)
+    {
+        String specificationVersion = connectorFactory.getClass().getPackage().getSpecificationVersion();
+        String implementationVersion = connectorFactory.getClass().getPackage().getImplementationVersion();
+        if (specificationVersion == null && implementationVersion == null) {
+            // The information comes from jar's Manifest and is not present e.g. when running tests, or from an IDE.
+            return Optional.empty();
+        }
+
+        requireNonNull(specificationVersion, "specificationVersion not present when implementationVersion is present");
+        requireNonNull(implementationVersion, "implementationVersion not present when specificationVersion is present");
+
+        // Implementation version comes from Manifest and is defined in Airbase as equivalent to `git describe` output.
+        if (implementationVersion.matches(".*-(\\d+)-g([0-9a-f]+)(-dirty)?$")) {
+            // Specification version comes from Manifest and is the project version with `-SNAPSHOT` stripped, and .0 added (if not present).
```

Comment:
magic... is that contract that will hold?

Reply:
https://github.com/airlift/airbase/blob/e58ac27e03696460f3cd9ec4400c4454f529121a/airbase/pom.xml#L512-L522